File: src/runtime/storageNG/drivers/volatile.ts
/**
 * @license
 * Copyright (c) 2019 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {StorageDriverProvider, DriverFactory} from './driver-factory.js';
import {Driver, ReceiveMethod, Exists} from './driver.js';
import {StorageKey} from '../storage-key.js';
import {Arc} from '../../arc.js';
import {ArcId} from '../../id.js';
import {RamDiskStorageKey} from './ramdisk.js';
import {Dictionary} from '../../hot.js';
import {assert} from '../../../platform/assert-web.js';
import {StorageKeyParser} from '../storage-key-parser.js';
import {Capabilities} from '../../capabilities.js';
import {CapabilitiesResolver, StorageKeyOptions} from '../../capabilities-resolver.js';

type VolatileEntry<Data> = {data: Data, version: number, drivers: VolatileDriver<Data>[]};
type VolatileEntryCollection<Data> = {root: VolatileEntry<Data>, locations: Dictionary<VolatileEntry<Data>>};

export class VolatileStorageKey extends StorageKey {
  public static readonly protocol = 'volatile';
  readonly arcId: ArcId;
  readonly unique: string;
  readonly path: string;

  constructor(arcId: ArcId, unique: string, path: string='') {
    super(VolatileStorageKey.protocol);
    this.arcId = arcId;
    this.unique = unique;
    this.path = path;
  }

  toString() {
    return `${this.protocol}://${this.arcId}/${this.unique}@${this.path}`;
  }

  childWithComponent(component: string) {
    return new VolatileStorageKey(this.arcId, this.unique, `${this.path}/${component}`);
  }

  // Note that subKeys lose path information.
  subKeyWithComponent(component: string) {
    return new VolatileStorageKey(this.arcId, `${this.unique}/${component}`);
  }

  static fromString(key: string): VolatileStorageKey {
    const match = key.match(/^volatile:\/\/([^/]+)\/([^@]*)@(.*)$/);
    if (!match) {
      throw new Error(`Not a valid VolatileStorageKey: ${key}.`);
    }
    const [_, arcId, unique, path] = match;
    return new VolatileStorageKey(ArcId.fromString(arcId), unique, path);
  }
}

export class VolatileMemory {
  entries = new Map<string, VolatileEntryCollection<unknown>>();
  // Tokens can't just be an incrementing number as VolatileMemory is the basis for RamDiskMemory too;
  // if we were to use numbers here then a RamDisk could be reaped, restarted, and end up with the
  // same token as a previous iteration.
  // When we want to support RamDisk fast-forwarding (e.g. by keeping a rotating window of recent
  // operations) then we'll need tokens to be a combination of a per-instance random value and a
  // per-operation updating number. For now, just a random value that is updated with each write
  // is sufficient.
  token = Math.random() + '';

  deserialize(data, unique: string) {
    assert(!this.entries.has(unique));
    const entry: VolatileEntryCollection<unknown> = {root: null, locations: {}};
    entry.root = {data: data.root, version: 0, drivers: []};
    if (data.locations) {
      for (const [key, value] of Object.entries(data.locations)) {
        entry.locations[key] = {data: value, version: 0, drivers: []};
      }
    }
    this.entries.set(unique, entry);
  }
}

/**
 * Allows for loosely coupled memory provisioning by clients of the storage
 * stack.
 */
export interface VolatileMemoryProvider {
  getVolatileMemory(): VolatileMemory;
}

export class SimpleVolatileMemoryProvider implements VolatileMemoryProvider {
  private readonly memory: VolatileMemory = new VolatileMemory();

  getVolatileMemory(): VolatileMemory {
    return this.memory;
  }
}

let id = 0;

export class VolatileDriver<Data> extends Driver<Data> {
  private memory: VolatileMemory;
  private pendingVersion = 0;
  private pendingModel: Data | null = null;
  private receiver: ReceiveMethod<Data>;
  private data: VolatileEntryCollection<Data>;
  private id: number;
  private path: string;

  constructor(storageKey: VolatileStorageKey | RamDiskStorageKey, exists: Exists, memory: VolatileMemory) {
    super(storageKey, exists);
    this.id = id++;
    this.memory = memory;
    this.path = null;
    if (storageKey instanceof VolatileStorageKey && storageKey.path !== '') {
      this.path = storageKey.path;
    }
    switch (exists) {
      case Exists.ShouldCreate:
        if (this.memory.entries.has(storageKey.unique)) {
          throw new Error(`requested creation of memory location ${storageKey} can't proceed as location already exists`);
        }
        this.data = {root: null, locations: {}};
        this.memory.entries.set(storageKey.unique, this.data as VolatileEntryCollection<unknown>);
        break;
      case Exists.ShouldExist:
        if (!this.memory.entries.has(storageKey.unique)) {
          throw new Error(`requested connection to memory location ${storageKey} can't proceed as location doesn't exist`);
        }
      /* falls through */
      case Exists.MayExist:
        {
          const data = this.memory.entries.get(storageKey.unique);
          if (data) {
            this.data = data as VolatileEntryCollection<Data>;
            this.pendingModel = this.localData();
            this.pendingVersion = this.localVersion();
          } else {
            this.data = {locations: {}, root: null};
            this.memory.entries.set(storageKey.unique, this.data as VolatileEntryCollection<unknown>);
            this.memory.token = Math.random() + '';
          }
          break;
        }
      default:
        throw new Error(`unknown Exists code ${exists}`);
    }
    this.pushLocalDriver(this);
  }

  private getOrCreateEntry() {
    if (this.path) {
      if (!this.data.locations[this.path]) {
        this.data.locations[this.path] = {data: null, version: 0, drivers: []};
      }
      return this.data.locations[this.path];
    }
    if (!this.data.root) {
      this.data.root = {data: null, version: 0, drivers: []};
    }
    return this.data.root;
  }

  private localData() {
    return this.getOrCreateEntry().data;
  }

  private localVersion() {
    return this.getOrCreateEntry().version;
  }

  private setLocalData(data: Data) {
    this.getOrCreateEntry().data = data;
  }

  private incrementLocalVersion() {
    this.getOrCreateEntry().version += 1;
  }

  private pushLocalDriver(driver: VolatileDriver<Data>) {
    this.getOrCreateEntry().drivers.push(driver);
  }

  registerReceiver(receiver: ReceiveMethod<Data>, token?: string) {
    this.receiver = receiver;
    if (this.pendingModel && token !== this.memory.token) {
      receiver(this.pendingModel, this.pendingVersion);
    }
    this.pendingModel = null;
  }

  getToken() { return this.memory.token; }

  async send(model: Data, version: number): Promise<boolean> {
    // This needs to contain an "empty" await, otherwise there's
    // a synchronous send / onReceive loop that can be established
    // between multiple Stores/Drivers writing to the same location.
    await 0;
    if (this.localVersion() !== version - 1) {
      return false;
    }
    this.setLocalData(model);
    this.incrementLocalVersion();
    this.getOrCreateEntry().drivers.forEach(driver => {
      if (driver === this) {
        return;
      }
      if (driver.receiver) {
        driver.receiver(model, this.localVersion());
      }
    });
    return true;
  }

  async write(key: StorageKey, value: Data): Promise<void> {
    throw new Error('Method not implemented.');
  }

  async read(key: StorageKey): Promise<Data> {
    throw new Error('Method not implemented.');
  }
}

/**
 * Provides Volatile storage drivers. Volatile storage is local to an individual
 * running Arc. It lives for as long as that Arc instance, and then gets
 * deleted when the Arc is stopped.
 */
export class VolatileStorageDriverProvider implements StorageDriverProvider {
  private readonly arc: Arc;

  constructor(arc: Arc) {
    this.arc = arc;
  }

  willSupport(storageKey: StorageKey): boolean {
    return storageKey.protocol === VolatileStorageKey.protocol
        && (storageKey as VolatileStorageKey).arcId.equal(this.arc.id);
  }

  async driver<Data>(storageKey: StorageKey, exists: Exists) {
    if (!this.willSupport(storageKey)) {
      throw new Error(`This provider does not support storageKey ${storageKey.toString()}`);
    }

    return new VolatileDriver<Data>(storageKey as VolatileStorageKey, exists, this.arc.volatileMemory);
  }

  // QUESTION: This method is never being called, is it needed?
  static register(arc: Arc) {
    DriverFactory.register(new VolatileStorageDriverProvider(arc));
  }
}

StorageKeyParser.addDefaultParser(VolatileStorageKey.protocol, VolatileStorageKey.fromString);
CapabilitiesResolver.registerDefaultKeyCreator(
    VolatileStorageKey.protocol,
    Capabilities.tiedToArc,
    (options: StorageKeyOptions) => new VolatileStorageKey(options.arcId, options.unique(), ''));
CapabilitiesResolver.registerDefaultKeyCreator(
    VolatileStorageKey.protocol,
    Capabilities.empty,
    (options: StorageKeyOptions) => new VolatileStorageKey(options.arcId, options.unique(), ''));

File: src_kt/java/arcs/storage/driver/Volatile.kt
/*
 * Copyright 2019 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.storage.driver

import arcs.common.ArcId
import arcs.common.toArcId
import arcs.storage.Driver
import arcs.storage.ExistenceCriteria
import arcs.storage.DriverFactory
import arcs.storage.DriverProvider
import arcs.storage.StorageKey
import arcs.storage.StorageKeyParser
import arcs.util.Random
import kotlinx.atomicfu.atomic
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** Protocol to be used with the volatile driver. */
const val VOLATILE_DRIVER_PROTOCOL = "volatile"

/** Storage key for a piece of data kept in the volatile driver. */
data class VolatileStorageKey(
  /** Id of the arc where this key was created. */
  val arcId: ArcId,
  /** Unique identifier for this particular key. */
  val unique: String
) : StorageKey(VOLATILE_DRIVER_PROTOCOL) {
  override fun toKeyString(): String = "$arcId/$unique"

  override fun childKeyWithComponent(component: String): StorageKey =
    VolatileStorageKey(arcId, "$unique/$component")

  override fun toString(): String = "$protocol://${toKeyString()}"

  companion object {
    private val VOLATILE_STORAGE_KEY_PATTERN = "".toRegex()

    init {
      // When VolatileStorageKey is imported, this will register its parser with the storage key
      // parsers.
      StorageKeyParser.addParser(VOLATILE_DRIVER_PROTOCOL, ::fromString)
    }

    private fun fromString(raw: String): VolatileStorageKey {
      val match =
        requireNotNull(VOLATILE_STORAGE_KEY_PATTERN.matchEntire(raw)) {
          "Not a valid VolatileStorageKey: $raw"
        }

      return VolatileStorageKey(match.groupValues[1].toArcId(), match.groupValues[2])
    }
  }
}

/** [DriverProvider] of [VolatileDriver]s for an arc. */
data class VolatileDriverProvider(private val arcId: ArcId) : DriverProvider {
  private val arcMemory = VolatileMemory()

  init {
    DriverFactory.register(this)
  }

  override fun willSupport(storageKey: StorageKey): Boolean =
    storageKey is VolatileStorageKey && storageKey.arcId == arcId

  override suspend fun <Data : Any> getDriver(
    storageKey: StorageKey,
    existenceCriteria: ExistenceCriteria
  ): Driver<Data> {
    require(willSupport(storageKey)) { "This provider does not support storageKey: $storageKey" }
    return VolatileDriver(storageKey, existenceCriteria, arcMemory)
  }
}

/** [Driver] implementation for an in-memory store of data. */
internal class VolatileDriver<Data : Any>(
  override val storageKey: StorageKey,
  override val existenceCriteria: ExistenceCriteria,
  private val memory: VolatileMemory
) : Driver<Data> {
  // The identifier is simply used to help differentiate between VolatileDrivers for the same
  // storage key.
  private val identifier = nextIdentifier.incrementAndGet()
  internal var receiver: (suspend (data: Data, version: Int) -> Unit)? = null
  private var pendingModel: Data? = null
  private var pendingVersion: Int = 0

  override val token: String?
    get() = memory.token

  init {
    require(
      // VolatileDriver does double-duty: serving both Volatile and RamDisk purposes, just with
      // different policies on which instances of VolatileMemory they point at.
      storageKey is VolatileStorageKey || storageKey is RamDiskStorageKey
    ) { "Invalid storage key type: $storageKey" }

    val dataForCriteria: VolatileEntry<Data> = when (existenceCriteria) {
      ExistenceCriteria.ShouldCreate -> {
        require(storageKey !in memory) {
          "Requested creation of memory location $storageKey can't proceed: already exists"
        }
        VolatileEntry()
      }
      ExistenceCriteria.ShouldExist ->
        requireNotNull(memory.get<Data>(storageKey)) {
          "Requested connection to memory location $storageKey can't proceed: doesn't exist"
        }.also {
          pendingModel = it.data
          pendingVersion = it.version
        }
      ExistenceCriteria.MayExist -> {
        val preExisting =
          memory.get<Data>(storageKey)?.also {
            pendingModel = it.data
            pendingVersion = it.version
          }

        // If we had a pre-existing value, return it. Otherwise make an empty one and put it in the
        // memory.
        preExisting ?: VolatileEntry()
      }
    }

    // Add the data to the memory.
    memory[storageKey] = dataForCriteria.copy(drivers = dataForCriteria.drivers + this)
  }

  override suspend fun registerReceiver(
    token: String?,
    receiver: suspend (data: Data, version: Int) -> Unit
  ) {
    this.receiver = receiver
    this.pendingModel
      ?.takeIf { this.token != token }
      ?.let { receiver(it, pendingVersion) }
    this.pendingModel = null
  }

  override suspend fun send(data: Data, version: Int): Boolean {
    val currentEntry: VolatileEntry<Data> = memory[storageKey]!!

    // If the new version isn't immediately after this one, return false.
    if (currentEntry.version != version - 1) return false

    val newEntry = VolatileEntry(data, version, currentEntry.drivers)
    memory[storageKey] = newEntry

    newEntry.drivers.forEach { driver ->
      val receiver = driver.takeIf { it != this }?.receiver
      receiver?.invoke(data, version)
    }

    return true
  }

  override fun toString(): String = "VolatileDriver(id=${identifier}, storageKey=$storageKey)"

  companion object {
    private var nextIdentifier = atomic(0)
  }
}

/** A single entry in a [VolatileDriver]. */
internal data class VolatileEntry<Data : Any>(
  val data: Data? = null,
  val version: Int = 0,
  val drivers: Set<VolatileDriver<Data>> = emptySet()
) {
  constructor(data: Data? = null, version: Int = 0, vararg drivers: VolatileDriver<Data>)
    : this(data, version, drivers.toSet())
}

/**
 * Lookup map of storage keys to entries, with a [token] that gets updated when data has changed.
 */
internal class VolatileMemory {
  private val lock = Any()
  private val entries = mutableMapOf<StorageKey, VolatileEntry<*>>()

  /** Current token. Will be updated with every call to [set]. */
  var token: String = Random.nextInt().toString()
    get() = synchronized(lock) { field }
    private set(value) = synchronized(lock) { field = value }

  /** Returns whether or not a [VolatileEntry] exists in memory for the [key]. */
  operator fun contains(key: StorageKey): Boolean = synchronized(lock) { key in entries }

  /**
   * Gets a [VolatileEntry] of type [Data] from the memory stored at [key], or null if not found.
   */
  @Suppress("UNCHECKED_CAST")
  operator fun <Data : Any> get(key: StorageKey): VolatileEntry<Data>? = synchronized(lock) {
    entries[key] as VolatileEntry<Data>?
  }

  /**
   * Sets the value at the provided [key] to the given [VolatileEntry] and returns the old value, if
   * a value was previously set.
   */
  @Suppress("UNCHECKED_CAST")
  operator fun <Data : Any> set(
    key: StorageKey,
    value: VolatileEntry<Data>
  ): VolatileEntry<Data>? = synchronized(lock) {
    val originalValue = entries[key]
    entries[key] = value
    token = Random.nextInt().toString()
    return originalValue as VolatileEntry<Data>?
  }

  /** Clears everything from storage. */
  fun clear() = entries.clear()
}


